File: frontend/src/types/appGraph.ts
import type {
  EntryActionCard,
  EntryGraphManifest,
  EntryTurnMessage,
  EntryUIArtifact,
  RouteDeckRuntimeSnapshot,
} from '@/types/entry'
import type { SaaSAgent } from '@/types/domain'

export interface AppGraphState {
  node: string
  active_saas_agent_id?: string | null
  active_connection_id?: string | null
  pending_trace_id?: string | null
  active_surface_id?: string | null
  route_params?: Record<string, unknown>
  navigation_back_stack?: Array<{
    node_id: string
    surface_id?: string | null
    params?: Record<string, unknown>
  }>
  navigation_forward_stack?: Array<{
    node_id: string
    surface_id?: string | null
    params?: Record<string, unknown>
  }>
  pending_operation_id?: string | null
  pending_operation_args?: Record<string, unknown>
  graph_context?: Record<string, unknown>
  executed_nodes?: string[]
}

export interface AppGraphContextLens {
  selected_saas_agent_id?: string | null
  selected_saas_agent_name?: string | null
  selected_saas_agent_slug?: string | null
  current_node: string
  working_on: string
  connection_count: number
  ready_connection_count: number
  action_count: number
  tool_count: number
  pending_trace_id?: string | null
  pending_trace_status?: string | null
}

export interface AppGraphResponse {
  state: AppGraphState
  graph_version: string
  graph_manifest: EntryGraphManifest
  route_deck_snapshot: RouteDeckRuntimeSnapshot
  context_lens: AppGraphContextLens
  available_actions: EntryActionCard[]
  persistent_actions: EntryActionCard[]
  ui_artifacts: EntryUIArtifact[]
  evidence: Record<string, unknown>[]
  diagnostics: Record<string, unknown>
  messages: EntryTurnMessage[]
  saas_agents: SaaSAgent[]
  replace_path?: string | null
}

File: backend/core/schemas/app_graph.py
from __future__ import annotations

import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field
from routedeck_core import (
    RouteDeckContextLens,
    RouteDeckGraphNavigationLocation,
    RouteDeckGraphRequest,
    RouteDeckGraphResponse,
    RouteDeckGraphState,
)

from .entry import (
    EntryActionCard,
    EntryGraphManifest,
    EntryGraphMessage,
    EntryRouteDeckRuntimeSnapshot,
    EntryUIArtifact,
)
from .saas_agent import SaaSAgentRead


class AppGraphNavigationLocation(RouteDeckGraphNavigationLocation):
    pass


class AppGraphState(RouteDeckGraphState):
    active_saas_agent_id: uuid.UUID | None = None
    active_connection_id: uuid.UUID | None = None
    pending_trace_id: uuid.UUID | None = None
    navigation_back_stack: list[AppGraphNavigationLocation] = Field(default_factory=list)
    navigation_forward_stack: list[AppGraphNavigationLocation] = Field(default_factory=list)


class AppGraphRequest(RouteDeckGraphRequest):
    state: AppGraphState | None = None
    saas_agent_id: uuid.UUID | None = None


class AppGraphRouterDecision(BaseModel):
    intent: Literal["action", "clarify", "no_match"] = "clarify"
    action_id: str | None = Field(default=None, max_length=160)
    node_id: str | None = Field(default=None, max_length=120)
    slots: dict[str, Any] = Field(default_factory=dict)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    clarification: str | None = Field(default=None, max_length=1_000)
    provider: str = "disabled"
    model: str | None = None


class AppGraphContextLens(RouteDeckContextLens):
    selected_saas_agent_id: uuid.UUID | None = None
    selected_saas_agent_name: str | None = None
    selected_saas_agent_slug: str | None = None
    connection_count: int = 0
    ready_connection_count: int = 0
    action_count: int = 0
    tool_count: int = 0
    router_index_status: str | None = None
    router_documents_count: int = 0
    router_endpoint_count: int = 0
    router_version: str | None = None
    pending_trace_id: uuid.UUID | None = None
    pending_trace_status: str | None = None


class AppGraphResponse(RouteDeckGraphResponse):
    state: AppGraphState
    graph_manifest: EntryGraphManifest
    route_deck_snapshot: EntryRouteDeckRuntimeSnapshot
    context_lens: AppGraphContextLens
    available_actions: list[EntryActionCard] = Field(default_factory=list)
    persistent_actions: list[EntryActionCard] = Field(default_factory=list)
    ui_artifacts: list[EntryUIArtifact] = Field(default_factory=list)
    messages: list[EntryGraphMessage] = Field(default_factory=list)
    saas_agents: list[SaaSAgentRead] = Field(default_factory=list)


class AppGraphTurnResponse(AppGraphResponse):
    turn_type: Literal["snapshot", "action", "turn"] = "turn"
